File: week05/32/241002.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head

        while (fast.next and fast.next.next):
            slow = slow.next
            fast = fast.next.next

        if fast.next == None:
            return slow

        else:
            return slow.next

File: week05/32/241002.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* middleNode(struct ListNode* head) {
    struct ListNode * slow = head;
    struct ListNode * fast = head;

    while (fast->next && fast->next->next) {
        slow = slow->next;
        fast = fast->next->next;
    }
    if (fast->next == NULL) {
        return slow;
    } else {
        return slow->next;
    }
}